### PR TITLE
Task Agnostic Metric Implementation

### LINKED ISSUE #10: Implement Task-agnostic metrics
- No. parameters (used in #4)
- No. layers (used in #6) - transformer equivalent probably no. blocks

### PR DESCRIPTION
**Pull request by eddableheath**
This pull request:

- Implements the task agnostic metric counting the parameters in a model.
- Adapts the locmoset/metrics/run.py script in order to run this metric.
- Implements the capacity to train metric scaling for #37 but does not incorporate it into the pipeline.
- Adds tests for the n_pars metric, as well as updating conftest.py.

### CODE REVIEW

File: src/locomoset/metrics/task_agnostic.py [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+Bolya, Daniel, Rohit Mittapalli, and Judy Hoffman. "Scalable diverse model selection
+for accessible transfer learning." Advances in Neural Information Processing Systems
+34 (2021): 19301-19312.
+"""
+
+import numpy as np
+from transformers import AutoModelForImageClassification
+
+
+def num_params_metric(model: AutoModelForImageClassification) -> int:
```
Comment (jack89roberts) [inline src/locomoset/metrics/task_agnostic.py]: I think we can be more generic in the type of model here, elsewhere we've used `transformers.modeling_utils.PreTrainedModel` (also `AutoModelForImageClassification` is more like a wrapper for getting a model rather than a model itself).

File: src/locomoset/metrics/task_agnostic.py [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        model: pre-loaded hugging face model from which to return the number of
+        parameters.
+
+    Returns:
+        number of paramters of said model.
+    """
+    return model.num_parameters()
+
+
+def rescale_by_params(models_and_metrics: dict) -> dict:
```
Comment (jack89roberts) [inline src/locomoset/metrics/task_agnostic.py]: Having a test for this would be good

File: src/locomoset/metrics/task_agnostic.py [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+
+    Args:
+        models_and_metrics: dictionary containing 'model_name': metric_value pairs.
+
+    Returns:
+        dictionary containgin 'model_name': scaled_metric_value paris.
+    """
+    n_pars = {}
+    for model in models_and_metrics.keys():
+        model_fn = AutoModelForImageClassification.from_pretrained(model)
```
Comment (jack89roberts) [inline src/locomoset/metrics/task_agnostic.py]: There's an interesting subtlety here - do we want the number of parameters including or excluding the classification head? My gut instinct is without the classification head, in which case the model should be loaded with `num_labels=0` here like we do elsewhere (but we should verify that actually changes the number of reported parameters).

This won't make (much of) a difference, especially if all the models we're using are pre-trained on ImageNet-1k, but if we start experimenting with models pre-trained on different datasets (with different numbers of classes) I think we're more interested in the number of parameters used to create the features and not the number of additional parameters in the final layer for classification on the pre-training dataset.

File: src/locomoset/metrics/task_agnostic.py (hunk 2 above)
Comment (eddableheath) [reply src/locomoset/metrics/task_agnostic.py]: A test for `rescale_by_params` in particular?

Comment (eddableheath) [reply src/locomoset/metrics/task_agnostic.py]: If so I hadn't written that yet as I hadn't considered it finished.

File: src/locomoset/metrics/task_agnostic.py (hunk 1 above)
Comment (eddableheath) [reply src/locomoset/metrics/task_agnostic.py]: Ok, will refactor that.

File: src/locomoset/metrics/task_agnostic.py (hunk 2 above)
Comment (jack89roberts) [reply src/locomoset/metrics/task_agnostic.py]: yeah for `rescale_by_params`, no probs 👍

File: src/locomoset/metrics/task_agnostic.py (hunk 3 above)
Comment (eddableheath) [reply src/locomoset/metrics/task_agnostic.py]: That definitely makes sense for the general case, also the number of parameters in the classification head shouldn't have a huge effect in the capacity to train for new information (I wouldn't have thought) but would skew the value of the metric.

File: src/locomoset/metrics/task_agnostic.py (hunk 1 above)
Comment (eddableheath) [reply src/locomoset/metrics/task_agnostic.py]: I've refactored this in [560fe72](https://github.com/alan-turing-institute/ARC-LoCoMoSeT/pull/47/commits/560fe7260eff453756ef47000d1db1797f1b1836) to use `transformers.modeling_utils.PreTrainedModel` as the type here and `transformers.AutoModel` for `rescale_by_params` for generality.

File: src/locomoset/metrics/task_agnostic.py (hunk 3 above)
Comment (eddableheath) [reply src/locomoset/metrics/task_agnostic.py]: Refactored in [560fe72](https://github.com/alan-turing-institute/ARC-LoCoMoSeT/pull/47/commits/560fe7260eff453756ef47000d1db1797f1b1836)

Comment (eddableheath) [reply src/locomoset/metrics/task_agnostic.py]: Also, to clarify, incorporating `num_labels=0` makes a big difference, for the `google/vit-base-patch16-384` model it's a difference of 769000 parameters.

**Comment by eddableheath** [comment]
This should be ready to go now.

File: src/locomoset/metrics/task_agnostic.py [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+Functions for the task-agnostic metrics. In particular the number of parameters of a
+model as a metric.
+
+This also includes functions for reweighting a set of metric values by the inclusion of
+a task agnostic metric as a proxy for 'capactity to learn', as introduced in the
+Pairwise Annotation Representation Comparison (PARC) metric for transferability paper:
+
+Bolya, Daniel, Rohit Mittapalli, and Judy Hoffman. "Scalable diverse model selection
+for accessible transfer learning." Advances in Neural Information Processing Systems
+34 (2021): 19301-19312.
```
Comment (jack89roberts) [inline src/locomoset/metrics/task_agnostic.py]: All of this can be removed now it's only the no. parameters metric I think:

```suggestion
```

File: tests/test_task_agnostic.py [hunk 5]
```diff
@@ -0,0 +1,12 @@
+"""
+Test functions for the task_agonstic.py metrics.
+"""
+
+from locomoset.metrics.task_agnostic import num_params_metric
+
+
+def test_num_params_metric(dummy_model_classifier):
```
Comment (jack89roberts) [inline tests/test_task_agnostic.py]: I'd be more inclined to use `dummy_model_head` here (or have a test with both showing that you get a different no. of parameters in each case), since we decided using the model head is probably more generalisable/predictive for what we're interested in.

File: src/locomoset/metrics/run.py [hunk 6]
```diff
@@ -104,6 +104,29 @@ def compute_metric(config: dict) -> dict:
     return results
 
 
+def compute_task_agnostic_metric(config: dict) -> dict:
```
Comment (jack89roberts) [inline src/locomoset/metrics/run.py:107]: We may need to revisit this and have a structure that makes it easier to abstract away different types of metrics, but as long as we only have two choices this works fine 👍

File: notebooks/ImageNet_check.ipynb
Comment (jack89roberts) [inline notebooks/ImageNet_check.ipynb:1]: could you check whether this notebook has anything worth keeping and either move it to the notebooks branch/delete it as appropriate

**Review by jack89roberts — CHANGES_REQUESTED** [review]

File: notebooks/ImageNet_check.ipynb
Comment (eddableheath) [reply notebooks/ImageNet_check.ipynb:1]: Done

File: tests/test_task_agnostic.py (hunk 5 above)
Comment (eddableheath) [reply tests/test_task_agnostic.py]: I ran some tests on this and they it does get different numbers of parameters, but not in a significant way across the models. I've now changed it to `dummy_model_head` and updated the test.

**Review by jack89roberts — APPROVED** [review]
